### PR TITLE
CART-963 build: Split Cart tests_common.h into .h and .c files

### PR DESCRIPTION
**Pull request by mallove79**
GitHub is having trouble rendering PR-5309. Port to new branch and squash commits.

### CODE REVIEW

File: src/utils/self_test/SConscript [hunk 1]
```diff
@@ -13,15 +13,18 @@ SELF_TEST = 'self_test.c'
 def scons():
     """scons function"""
 
-    Import('env', 'prereqs')
+    Import('env', 'prereqs', 'cart_lib', 'gurt_lib', 'cart_utils_objs')
 
     tenv = env.Clone(LIBS=[])
 
+    tenv.AppendUnique(CPPPATH=['#/src/cart', '#/src/tests/ftest/cart'])
```
Comment (jolivier23) [inline src/utils/self_test/SConscript:20]: What is ftest/cart needed for?

Also, src/cart is added in src/cart/SConscript so this shouldn't be necessary

File: src/utils/self_test/SConscript [hunk 2]
```diff
@@ -13,15 +13,18 @@ SELF_TEST = 'self_test.c'
...
     tenv.AppendUnique(LIBS=['cart', 'gurt', 'pthread', 'm'])
     prereqs.require(tenv, 'mercury')
+    tenv.AppendUnique(LIBPATH=['#/src/cart'])
```
Comment (jolivier23) [inline src/utils/self_test/SConscript:23]: similar comment to above

**Review by jolivier23 — CHANGES_REQUESTED** [review]

File: src/utils/self_test/SConscript (hunk 1 above)
Comment (mallove79) [reply src/utils/self_test/SConscript:20]: https://github.com/daos-stack/daos/pull/6115/commits/539a2817b782f8c63f13ec86d9b1c360ebd469e5

File: src/utils/self_test/SConscript (hunk 2 above)
Comment (mallove79) [reply src/utils/self_test/SConscript:23]: https://github.com/daos-stack/daos/pull/6115/commits/539a2817b782f8c63f13ec86d9b1c360ebd469e5

File: src/utils/self_test/SConscript (hunk 1 above)
Comment (mallove79) [reply src/utils/self_test/SConscript]: ~~539a281~~  77c5ef4018424f14f53546df1d488084cada1bd0

File: src/utils/self_test/SConscript (hunk 2 above)
Comment (mallove79) [reply src/utils/self_test/SConscript]: ~~539a281~~  77c5ef4018424f14f53546df1d488084cada1bd0

File: src/utils/ctl/SConscript [hunk 3]
```diff
@@ -12,13 +12,15 @@ SRC = ['cart_ctl.c']
 
 def scons():
     """scons function"""
-    Import('env', 'prereqs')
+    Import('env', 'prereqs', 'cart_lib', 'gurt_lib', 'cart_utils_objs')
 
     tenv = env.Clone(LIBS=[])
+    tenv.AppendUnique(CPPPATH=['#/src/cart', '#/src/tests/ftest/cart'])
```
Comment (jolivier23) [inline src/utils/ctl/SConscript:18]: missed this one

File: src/cart/SConscript [hunk 4]
```diff
@@ -71,12 +82,18 @@ def scons():
     env.AppendUnique(LINKFLAGS=["-Wl,-rpath-link=%s" % Dir('.').path])
     env.AppendUnique(CPPPATH=[Dir('.').srcnode()])
     env.AppendUnique(CPPPATH=[Dir('include').srcnode()])
+    env.AppendUnique(CPPPATH=[Dir('../gurt').srcnode()])
```
Comment (jolivier23) [inline src/cart/SConscript:85]: this should not be needed as well

File: src/tests/ftest/cart/SConscript [hunk 5]
```diff
@@ -62,43 +62,61 @@ FAULT_STATUS_TEST = 'fault_status.c'
...
-    Import('env', 'prereqs')
+    Import('env', 'prereqs', 'cart_lib', 'gurt_lib')
+    Import('cart_utils_objs')
 
+    ##############################
+    # Create test programs
+    ##############################
     tenv = env.Clone()
 
+    tenv.AppendUnique(CPPPATH=['#/src/cart/src/cart'])
```
Comment (jolivier23) [inline src/tests/ftest/cart/SConscript:73]: remove

File: src/tests/ftest/cart/SConscript [hunk 6]
```diff
@@ -62,43 +62,61 @@ FAULT_STATUS_TEST = 'fault_status.c'
...
     tenv.AppendUnique(LIBS=['cart', 'gurt', 'pthread', 'm'])
     prereqs.require(tenv, 'crypto', 'mercury')
+    tenv.AppendUnique(LIBPATH=[Dir('.')])
```
Comment (jolivier23) [inline src/tests/ftest/cart/SConscript:76]: should not be needed

File: src/tests/ftest/cart/SConscript [hunk 7]
```diff
@@ -62,43 +62,61 @@ FAULT_STATUS_TEST = 'fault_status.c'
...
     for test in TEST_GROUP_NP_TESTS:
-        target = daos_build.test(tenv, test, install_off='../../../../')
+        target = daos_build.test(tenv, [test, cart_utils_objs],
+                                 install_off='../../../../')
+        tenv.Requires(target, [cart_lib, gurt_lib])
         tenv.Install(tests_dir, target)
 
     # Build FAULT_STATUS_TEST separately and install it into bin/
     target = daos_build.test(tenv, FAULT_STATUS_TEST)
+    tenv.Requires(target, [cart_lib, gurt_lib])
```
Comment (jolivier23) [inline src/tests/ftest/cart/SConscript:112]: Not sure these Requires lines are necessary

File: src/utils/ctl/SConscript [hunk 8]
```diff
@@ -12,13 +12,15 @@ SRC = ['cart_ctl.c']
...
     tenv.AppendUnique(LIBS=['cart', 'gurt', 'pthread'])
     prereqs.require(tenv, 'mercury')
 
-    ctl = daos_build.program(tenv, SRC)
+    ctl = daos_build.program(tenv, [SRC, cart_utils_objs])
+    tenv.Requires(ctl, [cart_lib, gurt_lib, cart_utils_objs])
```
Comment (jolivier23) [inline src/utils/ctl/SConscript:23]: another Requires line

File: src/utils/self_test/SConscript [hunk 9]
```diff
@@ -13,15 +13,16 @@ SELF_TEST = 'self_test.c'
...
 
     tenv = env.Clone(LIBS=[])
 
     tenv.AppendUnique(LIBS=['cart', 'gurt', 'pthread', 'm'])
     prereqs.require(tenv, 'mercury')
     tenv.AppendUnique(FLAGS='-pthread')
 
-    self_test = daos_build.program(tenv, SELF_TEST)
+    self_test = daos_build.program(tenv, [SELF_TEST, cart_utils_objs])
+    tenv.Requires(self_test, [cart_lib, gurt_lib])
```
Comment (jolivier23) [inline src/utils/self_test/SConscript:25]: another requires line.   I would test fix all of the CPPPATH and LIBPATH issues, make a commit, then remove these and see if build still works...  If not, defer cleanup or Requires to another patch.

**Review by jolivier23 — CHANGES_REQUESTED** [review]
Sorry, missed a few in prior review

File: src/utils/self_test/SConscript (hunk 9 above)
Comment (mallove79) [reply src/utils/self_test/SConscript:25]: CI seems to [build/run okay](https://build.hpdd.intel.com/blue/organizations/jenkins/daos-stack%2Fdaos/detail/PR-6115/4) [without the Requires](https://github.com/daos-stack/daos/pull/6115/commits/95f424939288f9a595a9ffc5a25ab39b5ef1b0e5).

File: src/tests/ftest/cart/SConscript [hunk 10]
```diff
@@ -98,7 +107,8 @@ def scons():
     benv = tenv.Clone()
 
     benv.AppendUnique(CFLAGS=['-std=gnu99', '-pedantic'])
-    daos_build.test(tenv, BASIC_SRC)
+    basic_target = daos_build.test(tenv, BASIC_SRC)
+    benv.Requires(basic_target, [cart_utils_objs])
```
Comment (jolivier23) [inline src/tests/ftest/cart/SConscript:111]: missed one here...maybe can clean up later

**Review by jolivier23 — DISMISSED** [review]

**Review by frostedcmos — DISMISSED** [review]

**Review by frostedcmos — APPROVED** [review]

**Review by jolivier23 — APPROVED** [review]

**Pull request merged**